### ISSUE TITLE
recycling x and y inefficient

### ISSUE THREAD

**Issue by edzer** [r-spatial/s2#34]
We now compute pairwise metrics for `x` and `y` by recycling them, looping over all the pairs, and carrying out the entire operation for each pair individually. This is advantageous when
* `x` and `y` both have a single feature
* `x` and `y` are sets of equal size, without repetition.
A very common use case though (I believe) is that `x` is very large, and `y` has size one. E.g.

* compute the nearest feature from a set for 100.000 addresses
* compute distances to the set of fire polygons for all raster cells
* find out which cars are on a highway, and which don't

I believe (without having tested) that we would get tremendous speed increase when we would process cases with single `y` specially: the index structure is then built only once, and queried many times. This is similar to building the `STRtree` in GEOS, which gave a massive speed boost. To get there, we shouldn't recycle `y` in the first place.

**Comment by paleolimbot** [comment]
Definitely test first....it's the pointer that's being recycled and the underlying `LibS2Geography` has lazy index building already!

**Comment by edzer** [comment]
Oh that's really cool - hadn't given that a single thought!!

**Comment by paleolimbot** [comment]
I think it's too soon to optimize here, and based on my limited benchmarking, I don't think recycling or index construction are the first problem...the main speed hit is actually constructing the objects. I don't think this is a wk package problem based on benchmarking that against sf's WKB IO (see the readme)...hard to tell just yet. Closing this in favour of #44...
